specs/trip_dispatch_spec.rb

```diff
@@ -111,4 +108,90 @@
       expect(passenger.trips).must_include trip
     end
   end
+
+  describe "request_trip method" do
+    before do
+      @dispatcher = RideShare::TripDispatcher.new(USER_TEST_FILE, TRIP_TEST_FILE, DRIVER_TEST_FILE)
+
+      @before_trips = @dispatcher.drivers[2].driven_trips.length
+
+      @before_new_trip_driver = @dispatcher.drivers[2].status
+
+      @before_passenger_trips = @dispatcher.passengers[0].trips.length
+
+      @dispatcher.request_trip(1)
+
+
+    end
+
+    it "creates new instance of trip" do
+      expect(@dispatcher.trips.last).must_be_kind_of RideShare::Trip
+    end
+
+    it "assigns first :AVAILABLE driver to the ride" do
+      expect(@dispatcher.trips.last.driver.id).must_equal 6
+    end
+
+    it "assigns an available driver to the ride" do
+      expect(@before_new_trip_driver).must_equal :AVAILABLE
+    end
+
+    it "should start time now" do
+      expect(@dispatcher.trips.last.start_time).must_be_close_to Time.now
+    end
+
+    it "should have an endtime of nil" do
+      expect(@dispatcher.trips.last.end_time).must_equal nil
+    end
+
+    it "should have a rating of nil" do
+      expect(@dispatcher.trips.last.rating).must_equal nil
+    end
+
+    it "adds new instance of requested trip to driver's driven_trips array" do
+
+      expect(@dispatcher.drivers[2].driven_trips.length).must_equal (@before_trips + 1)
+
+      expect(@dispatcher.drivers[2].driven_trips.last.passenger.id).must_equal 1
+    end
+
+    it "adds new instance of requested trip to passengers's trips array" do
+
+      expect(@dispatcher.passengers[0].trips.length).must_equal (@before_passenger_trips + 1)
+
+      expect(@dispatcher.passengers[0].trips.last.driver.id).must_equal 6
+    end
+
+    it "changes drivers status to unavailble after assigning to requested trip" do
+
+      expect(@dispatcher.drivers[2].status).must_equal :UNAVAILABLE
+    end
+
+    it "total revenue doesn't change after in progress trip added" do
+      expect(@dispatcher.drivers[1].total_revenue).must_equal 40.04
+      @dispatcher.request_trip(1)
+      expect(@dispatcher.drivers[1].total_revenue).must_equal 40.04
+    end
+
+    it "raises an error if the driver tries to drive themselves" do
+      @dispatcher.request_trip(2)
+      expect{@dispatcher.request_trip(5)}.must_raise ArgumentError
+    end
+
+    it "has a preference for drivers with least amount of trips" do
+
+      test_trip_input = {:id => 1234, :passenger => "George Foreman", :end_time => Time.parse("2018-05-25 12:25:00 -0700"), :start_time => Time.parse("2018-05-25 11:52:40 -0700"), :cost => 100, :rating => 3}
+
+      test_ride = RideShare::Trip.new(test_trip_input)
+
+
+      @dispatcher.drivers[0].status = :AVAILABLE
+      @dispatcher.drivers[2].add_driven_trip(test_ride)
+
+      expect(@dispatcher.request_trip(1).driver.id).must_equal 7
+
```

Comment:
What if there is no available driver? This is an important case, and is exactly the sort of situation where software will do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  @dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    @dispatcher.request_trip(2)
  }.must_raise StandardError
end
```

Reply:
dang! we put an argumenterror in trip_dispatcher.rb line 148 but must've forgotten to write a test.